Introduce new `unmanaged-cluster` model

joshrosso opened the pull request:
## What this PR does / why we need it

See [this page to test this PR and understand details around its design](https://github.com/vmware-tanzu/community-edition/blob/standalone-overhaul/cli/cmd/plugin/standalone-cluster/README.md).

Please [add all feedback around this proposal to the proposal issue](https://github.com/vmware-tanzu/community-edition/issues/2266).

Please keep this PR exclusive to feedback about code.

Fixes: https://github.com/vmware-tanzu/community-edition/issues/2266

Commit:
Initial commit for local clusters

This commit provides local cluster functionality to the tanzu-cli. It
also replaces the existing standalone-cluster plugin. As of this commit
it does work, but makes several assumptions about the environment.

jpmcb committed:
Implements pulling kapp image bundle and parsing it with ytt

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Remove ethtool containers when done

During cluster deployment, the ethtool container is run a few times to
get network information. These containers are run by calling out to
docker run, then they are left behind after the output is received from
them.

This adds the `--rm` argument to the `docker run` command so these
containers go away after we are done with them.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Add cluster module to local plugin

This adds an abstraction for working with clusters to the local plugin.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Implements the TkrImageReader interface

enables operations on an individual bundle image.

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Implement kapp package

This package is reposible for installing and providing the status of
kapp-controller

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Start to refactor kind calls into interface

This moves some of the cluster interface into a kind specific
implementation.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Add ability to list nodes

This completes removal of direct kind interaction from the command code
and routes everything through the cluster manager abstraction.

Also includes minor cleanup.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Various code cleanup

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Start of a cluster configuration API (#2279)

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Implements adding values files for a image bundle (#2283)

- Bumps ytt to v0.37.0
- Implements AddYttYamlValuesBytes method to TkrImageReader
- Implements AddYttKVsFromYAML method to TkrImageReader
- Uses a null logger for imgpkg so logging is surpressed

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
introduce tanzu package (#2287)

* move scheme resolution into kappmanager

this prevents creating the scheme for every instance of object detection
we need to do.

* Introduce the tanzu package

This package is responsible for orchestrating all the paces from tkr to
cluster creation to package install.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Remote path-node-for-antrea shell script (#2285)

This removes the platform-dependent shell script we were using to patch
the node networking for Antrea. These changes are now called from the go
code, making it possible to run on any platform with a docker
executable.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
remove dead code from old create command and remove duplicate antrea patch

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Move antrea network patching to kind create (#2290)

This moves the step for patching the container networks to happen during
creation of the cluster. This keeps low level steps together as part of
that process and will make it easier to add conditional handling once we
start working with other CNIs and providers.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Small log update to mention installing kapp-controller

Signed-off-by: joshrosso <rossoj@vmware.com>

jpmcb committed:
Implements getting TKR bom from registry (#2293)

- Implements DownloadImage for TkrImageReader
- Tanzu package now has getTkrBom which uses TkrImageReader
  to pull in the bom image file

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Make binary portable. (#2294)

No longer rely on configuration folders inside the code repo.

There are some consts introduced that are not intended for long term.

Signed-off-by: joshrosso <rossoj@vmware.com>

jpmcb committed:
Implements using ~/.config/tanzu/tkg/local as the config dir (#2305)

- Adds `getLocalBomPath()` function
- Renames `bundle` to `bomImage` in `getTkrBom` function

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Wire in config struct for local cluster create (#2307)

This starts the process for adding the use of a config struct for
managing all of the local cluster configuration settings. For now it
just hardwires it in to the creation call, but will be expanded to be
something passed in with different settings that can then be used to
change how we create the cluster.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Use io.Copy instead of os.Rename for bom downloading (#2310)

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Introduce package detection (#2311)

This looks up available packages to determine whether the CNI can be
resolved from a version and fqName perspective.

jpmcb committed:
Implements generating bom filename for local bom config dif (#2314)

- Adds logging events for bom downloading / detection

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Plumb local cluster create configuration (#2317)

This starts the implementation of getting the configuration from the
command line for creating a new local cluster and using that during the
deployment process.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
small changes to config for local cluster (#2319)

Signed-off-by: joshrosso <rossoj@vmware.com>

Commit:
Remove github.com/ghodss/yaml dependency in favor of gopkg.in/yaml.v3 (#2324)

Introduce package defaults and advance config

Our import of ghodss/yaml was to support a very small function which
wrapped gopkg.in/yaml.v3. Rather than making our dependency graph more
complicated, the project now import gopkg.in/yaml.v3 exclusively.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Load local cluster config by order of precedent (#2332)

This updates how we determine local cluster configuration by setting the
values based on:

1. explicit arguments
2. environment variables
3. configuration files
4. default values

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
introduce config pkg and plumb config (#2333)

With configuration now flowing through local, this commit forwards the
configuration from the tanzu package to the various packages it
interacts with.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Fix local cluster config unit tests (#2335)

Variable rename broke unit tests. This updates the variable.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
fail when CNI cannot be resolved

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Fix lint issues and cleanup (#2336)

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Co-authored-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Add CLI args for local cluster create and configure (#2337)

This adds the remaining CLI arguments to the local cluster create and
configure commands to be able to provide the most of the configuration
to use as part of the command line.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Introduce config directory per cluster (#2344)

This introduces a configuration directory per cluster at
`~/.config/tanzu/tkg/local/${CLUSTER_NAME}`. It stores the rendered
(flattened) configuration there.

This commit also re-wires the delete command to support multiple
provides and load the config.

Signed-off-by: joshrosso <rossoj@vmware.com>

jpmcb committed:
Implements `repository` field on all TKR components (#2343)

- Refactors the tkr.go file to have fewer functions
- Adds get methods for node and kapp images to get their custom
repository

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Address local cluster lint issues (#2346)

Cleans up a few issues that we pushed in.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Handle empty CLI default config values (#2347)

Initializing the local cluster configuration was checking for the
presence of config settings, but not whether they actually contained a
value. This is a problem with the real CLI processing since the default
values set on all options ends up being "", resulting in default values
being used unless passed in as a command line argument.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Add missing `Repository` for CSI elements in TKR (#2348)

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Inline some TKR info (#2349)

This tries to keep things DRY and hopefully gets a step closer to having
a simplified TKR structure.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Rename local plugin to standalone. (#2358)

This also revamps the description in the binary.

Signed-off-by: joshrosso <rossoj@vmware.com>

jpmcb committed:
Implements using node image from TKR (#2350)

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Set default tty mode based on runtime (#2345)

This sets the default value for the `--tty` CLI setting for local
cluster create and configure to match what the current runtime supports.
This makes it so the user shouldn't need to specify the `--tty` flag in
most cases unless they have a specific reason for changing it.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Update ytt dependency to be portable on Windows. (#2368)

Due to an issue described in
https://github.com/vmware-tanzu/carvel-ytt/issues/524, we need to rely
on a fork of `ytt` which changes an expected constant/key. Once this is
closed, we should go back to pointing at an official tag. Without this
in place, ytt parsing will fail when our binary is run on Windows.

Signed-off-by: joshrosso <rossoj@vmware.com>

Commit:
Standalone README (#2360)

* Add README detailing the standalone plugin

Signed-off-by: joshrosso <rossoj@vmware.com>

Commit:
add chmod to readme

Signed-off-by: joshrosso <rossoj@vmware.com>

Commit:
fix cp command in readme

Commit:
Switch to the TCE TKR-BOM (#2369)

This introduces our own custom TKR-BOM as the default. The primary
changes is this uses our kind image, compatible with the newest version
of kind and newest version of k8s 1.21.x.

Signed-off-by: joshrosso <rossoj@vmware.com>

Commit:
update note on create

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Improve TTY detection for Windows terminals (#2380)

The newer Windows Terminal supports unicode, but cmd.exe and powershell
do not. This improves our TTY detection to recognize if we are running
on Windows, and if so, defaults to TTY support only if it looks like we
are running in Windows Terminal.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Implements the noop cluster manager (#2375)

- the Kubernetes cluster interface now uses loaded byte slice instead
  of a path for the kubeconfig. Kapp manager changed accordingly.
- Removes list from cluster manager
- Enables users to bring their own cluster with their own kubeconfig
- Adds error logging in kapp controller manager
- Updates calls to NewKindClusterManager to be generic

Signed-off-by: John McBride <jmcbride@vmware.com>

davidvonthenen committed:
Fix release packaging (#2399)

stmcginnis committed:
Send standalone bootstrap log to file (#2402)

This updates the logging for the local cluster creation process to
capture all activity to the bootstrap.log file in the cluster's config
directory.

Signed-off-by: Sean McGinnis <sean.mcginnis@gmail.com>

jpmcb committed:
Renames `local` to `standalone` within codebase (#2417)

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb committed:
Implements a set of standard 2 width emojis for logging (#2420)

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Table format standalone cluster list (#2421)

This updates the list output for standalone clusters to show the name
and provider in a table format.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Separate image pull from cluster creation (#2408)

In regions where the cluster image hasn't been cached yet, it can take
some time to pull down the image. Right now this just happens during
cluster creation, so the CLI prints out "Creating cluster", then appears
to hang for a very long time.

We can't do too much to speed up the image pull, but we can at least
break it out into its own step so it's more obvious to the user what is
happening. This adds a separate step for getting the image first before
continuing on to the cluster creation.

Also cleans up some log handling while we're at it.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Replaces ColorLiteGrey with ColorBrightBlack (#2441)

- uses code 90 bright black in support of lite theme terminals

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Use client-go rest.Inteface for all packaging interactions (#2426)

By using rest.Interface, we won't be in a place where client-go versions
are broken between us and our dependencies, in this case,
kapp-controller.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Plumb standalone cluster network arguments (#2446)

This plumbs through the --pod-cidr and --service-cidr CLI arguments to
kind cluster creation. Input validation still needs to be done, but
assuming the correct CIDR format is used, clusters will be set up with
the corresponding pod and service subnets.

This also gets things in place to be able to perform other future kind
configuration customizations by populating the kind cluster struct
rather than parsing a static string.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Add port mapping to sa cluster creation (#2451)

This adds the ability to provide a list of ports to map from container
to host when creating a new standalone cluster. Multiple port mappings
can be provided by specifying the --port-map or -p switch and giving it
in the Docker format of container:host/protocol.

Note: the -p switch had previously been used as the short form for
--provider. This removes that and makes -p the short form for --port-map
since that is more consistent with how tools like the docker CLI handle
mapping or publishing ports.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Implements go routines for animating progress (#2401)

- AnimateProgressWithOptions takes AnimatorOptions
  to display a message with a ticking animation

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Clean up standalone packages and commands (#2450)

Signed-off-by: joshrosso <rossoj@vmware.com>

jpmcb committed:
Fixes 1 column emoji in delete command and fixes tty animating (#2473)

- Replaces "SawEmoji" with "TestTubeEmoji" which is 2 colums
- Adds newline to count 0 for tty logging in animating progress

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb committed:
Updates standalone-overhaul readme with updated video (#2475)

* Use ascii cast and update standalone readme

Signed-off-by: John McBride <jmcbride@vmware.com>

* Update README.md

* Update README.md

stmcginnis committed:
Rename new standalone to unmanaged (#2763)

This renames the new implementation for standalone clusters to be
unmanaged-clusters.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Restore existing standalone-cluster plugin (#2776)

When we started working on the replacement for the standalone-cluster
plugin, we made the assumption that it would be an in-place replacement.
Based on that, the new implementation was done by modifying the existing
code.

We now need to support a deprecation period for standalone-cluster, with
the new implementation being made available as the unmanaged-cluster
plugin. In our feature development branch we have renamed the modified
standalone-cluster plugin directory to its new name of
unamanged-cluster. This restores the existing standalone-cluster plugin
source so when we merge the feature branch to the main branch we don't
lose the standalone-cluster plugin.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/main.go:
```diff
@@ -0,0 +1,51 @@
+
+var descriptor = cliv1alpha1.PluginDescriptor{
+	Name:        "unmanaged-cluster",
+	Aliases:     []string{"um", "un", "unmanaged"},
```
@stmcginnis @jpmcb I'd like to propose we make our aliases:

* `uc`
* `unmanaged`

Does that seem sensible to ya'll?

stmcginnis replied:
Those looks good, but I would also argue for `um`. That would be my default assumption based on the name.

jpmcb replied:
I agree with Sean:
- `um`
- `uc`
- `unmanaged`

joshrosso replied:
sgtm, can someone make this change in a PR?

stmcginnis replied:
On it.

joshrosso commented:
@stmcginnis The tab writer we brought in from framework, it bolds the column headers of output, which I don't believe we should have for this plugin. While it is possible to work around this with OS detection, it still causes issue in many context such as using tools like `watch` where you get:

![image](https://user-images.githubusercontent.com/6200057/148793217-f4d517bd-1693-4042-8f38-ceb93b0b22ed.png)

Are you or @jpmcb able to remedy this? I'm perfectly fine with copying the code over from framework.

stmcginnis commented:
Yep, I'll address that today.

joshrosso commented:
We're getting close, so marking this as ready for review. A few known things that need to happen:

* [ ] @stmcginnis, @joshrosso, @jpmcb go through another full review
* [ ] @joshrosso communicates the proposal is accepted to the mailing list
* [ ] existing comments/requests for changes are addressed

Documentation is also a pre-req for release in v0.10.0, but I want to keep that to a separate issue/PR.

Commit:
unmanaged-cluster: remove usage of grey log text (#2791)

This commit removes the usage of grey log text in favor of no text
stylization in bootstrapping.

This approach introduces ColorNone rather than trying to stylize under
a white value. This is because many modern terminal emulators let you
overwrite the meaning of white. This will result in the terminal's
default "no style". ColorBrightBlack has been removed since it is no
longer used; it can be re-introduces should it be used in the future.

This usage of they grey text was intended to focus the eye on
Event/Eventf commands. However, based on terminal stylization, we found
that this darker text had a high-variance in visibility. Atop that, we
should do more research in the realm of accessibility before dimming out
such text as a default value.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Set unmanaged-cluster aliases to agreed list (#2792)

Based on conversations, we've decided to set the unmanaged-cluster
aliases to:

- um
- uc
- unmanaged

This is a minor fix to update that list of aliases that are provided on
plugin registration.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb commented:
@dvonthenen - looks like there's a merge conflict in the `hack/release/package-release.sh` script. My gut feeling is that chunk just needs to be removed as it was moved lower, but want to confirm with you

joshrosso commented on cli/cmd/plugin/standalone-cluster/create.go:
```diff
@@ -43,9 +45,17 @@ func init() {
 }
 
 func create(cmd *cobra.Command, args []string) error {
+	fmt.Print("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
```
@jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

jpmcb replied:
@joshrosso - how's this wording look?

![Screen Shot 2022-01-10 at 11 06 07 AM](https://user-images.githubusercontent.com/23109390/148816372-1ad9170c-efbc-4b11-a4f3-2900c445eacd.png)

jpmcb replied:
FYI - https://github.com/vmware-tanzu/community-edition/pull/2795

joshrosso replied:
love it. going to make one small recommendation that I missed (will do it in that PR.

joshrosso commented on cli/cmd/plugin/standalone-cluster/delete.go:
```diff
@@ -43,6 +43,13 @@ func init() {
 }
 
 func teardown(cmd *cobra.Command, args []string) error {
+	fmt.Print("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
```
@jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

joshrosso commented on cli/cmd/plugin/standalone-cluster/main.go:
```diff
@@ -12,9 +12,23 @@ import (
-	Group:       cliv1alpha1.RunCmdGroup,
+	Name: "standalone-cluster",
+	Description: `
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
@jpmcb can you update this deprecation notice updated to mention the use of `unmanaged-cluster` since it will ship alongside `standalone-cluster`

jpmcb committed:
Updates deprecation notice in standalone-cluster (#2795)

- Mentions the new unmanaged-cluster plugin which will live side by side
  with standalone
- Gives simple example of how to use `unmanaged-cluster`

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb committed:
Standalone deprecation notice update (#2797)

Now states that standalone _is_ deprecated.

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb commented on cli/cmd/plugin/unmanaged-cluster/README.md:
```diff
@@ -0,0 +1,402 @@
+
+1. Download the (unsigned) binary.
+
+    * [Linux](https://storage.googleapis.com/tce-tanzu-cli-plugins/build-tools/unmanaged/linux/tanzu-unmanaged-cluster)
```
These Google storage links appear broken and failing now in the URL linter. When I click the link, I see:

![Screen Shot 2022-01-10 at 3 12 36 PM](https://user-images.githubusercontent.com/23109390/148847373-b19899de-dae9-4b10-9883-dff12cfeed73.png)

jpmcb replied:
The binaries appear to still be at:

https://storage.googleapis.com/tce-tanzu-cli-plugins/build-tools/standalone/linux/tanzu-standalone

do we want to build some new ones and put them in the right bucket?

joshrosso replied:
> Or should we remove this chunk in the README? We probably don't need a pre-built binary of it if it's in the main release and on the main branch

My vote is to remove it in the README. Totally agreed that the pre-built binary isn't required and since the RFC window has close, folks can either:

1. get it in the v0.10.0 release
2. build from source

thoughts?

jpmcb committed:
Updates standalone-cluster description with short description notice (#2801)

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb committed:
Add plugin descriptor version number using buildinfo.Version (#2800)

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
unmanaged-cluster: add line break after error output (#2839)

This inserts a line break when an error is returned to the calling
client (cmd packag). It improves readability as PS1 of the terminal will
not start at the end of the error message.

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Add preflight checks to cluster deployment (#2817)

This adds a PreflightCheck call to the cluster managers to be able to
perform some checks up front for issues that will cause problems later
on. For the kind provider, this entails checking presence of docker and
the CPU and memory available to the docker engine.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Address lint issues (#2829)

With the merge into this development branch from main, we've picked up
the latest golangci-lint settings. These point out a few issues that
were not found with the older version of the linters. We've also been
ignoring a few things, but since we've switched from golint to revive
for some checks they were being highlighted again.

To address new issues and to stop ignoring others, this makes various
minor updates to avoid linting complaints.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Standalone plugin cobra command name should not have dep notice (#2849)

- The deprecation notice for standalone cluster should only be in the
  description. This removes the deprecation notice for standalone
  clusters from it's name in the cobra command. This is in support of
  `tanzu plugin list` and calling the command via the `tanzu` higher
  order command.

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Add output formats to unmanaged-cluster list (#2840)

This ties in the ability to output in yaml, json, and the default table
format.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Fork output writer to avoid bold table headers (#2841)

We had been using the OutputWriter from tanzu-framework. In the release
we are using, the header values for table columns in that version would
bold the text, resulting in garbage characters when doing things like
`watch` to wait for conditions in the output. This has since been fixed,
but is not available in a tanzu-framework release we can use yet.

Until we are able to bump up to a version of tanzu-framework that has
the fix, this temporarily duplicates that code locally. This can be
removed in a future release.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Introduce go color library with minimal dependencies (#2822)

- The keeps us from having to manage escape codes
and also gives us the advantage of bailing out on the use of colors when
$NO_COLOR is present in the env. Works with windows terminals,
and across many different terminal themes.

- Replaces "successful" info messages with faded messages to indicate
a success.

Signed-off-by: John McBride <jmcbride@vmware.com>

stmcginnis committed:
Change tty control flag to --tty-disable (#2850)

Bool flags for cobra commands are defaulted to true when present, and
the syntax of needing an = to set the value to something other than that
isn't the most intuitive.

Rather than have our tty control flag default to true, requiring the
user to explicitly set the value to switch it to false, this simplifies
things by changing the flag to `--tty-disable`. That makes its usage a
little simpler in that when it is actually used, the intent of using it
is to disable the TTY formatting in our output. So now it can just be
used as a simple flag.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

joshrosso commented:
@stmcginnis @jpmcb this is officially open for review.

At a minimum, can you both build and validate:

* cluster creation
* cluster configuration
* cluster creation with configuration (some minimal settings tweaked)
* cluster deletion
* cluster listing

based on results of the above and/or any code review please:

* block and request changes
* approve for merge

Lets get 2 approvals before merge.

joshrosso commented:
The following testing was performed after running `make install-all-tanzu-cli-plugins ENVS=linux-amd64` off 9cf7e3cac08b91997d1e031033547ec9492fad48. I am running on a system using cgroupsv2

### Created vanilla cluster :heavy_check_mark: 

```
$ tanzu unmanaged-cluster create doggo

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   Downloaded to: /home/josh/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/josh/.config/tanzu/tkg/unmanaged/doggo/config.yaml
   Bootstrap Logs: /home/josh/.config/tanzu/tkg/unmanaged/doggo/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster doggo
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/josh/.config/tanzu/tkg/unmanaged/doggo/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to doggo

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete doggo
```

```
$ k get po -A
NAMESPACE            NAME                                          READY   STATUS    RESTARTS   AGE
kube-system          antrea-agent-jp5sk                            2/2     Running   0          75s
kube-system          antrea-controller-68b8dfb8d6-tdm5k            1/1     Running   0          74s
kube-system          coredns-558bd4d5db-tt7xg                      1/1     Running   0          2m16s
kube-system          coredns-558bd4d5db-v5jrv                      1/1     Running   0          2m16s
kube-system          etcd-doggo-control-plane                      1/1     Running   0          2m20s
kube-system          kube-apiserver-doggo-control-plane            1/1     Running   0          2m20s
kube-system          kube-controller-manager-doggo-control-plane   1/1     Running   0          2m20s
kube-system          kube-proxy-rcmv5                              1/1     Running   0          2m16s
kube-system          kube-scheduler-doggo-control-plane            1/1     Running   0          2m20s
local-path-storage   local-path-provisioner-547f784dff-cz8vw       1/1     Running   0          2m16s
tkg-system           kapp-controller-f7dbbcb4f-n2d9x               1/1     Running   0          2m16s

```

### Create configuration file :heavy_check_mark: 

```
$ tanzu uc configure newtest
Wrote configuration file to: newtest.yaml
```

```
$ cat newtest.yaml
ClusterName: newtest
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: antrea
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false
```

### list clusters :heavy_check_mark: 

```
$ tanzu uc ls
  NAME   PROVIDER
  doggo  kind
```

### delete clusters with tty off :heavy_check_mark: 

```
$ tanzu uc rm doggo --tty-disable

 Deleting cluster: doggo

 Deleted cluster: doggo
```

### deployed cluster with calico set as cni in config

```
$ cat newtest.yaml
ClusterName: boooooop
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false
```

note `boooooop`

note `cluster name` is boooop and calico is selected

```
📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/josh/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/josh/.config/tanzu/tkg/unmanaged/boooooop/config.yaml
   Bootstrap Logs: /home/josh/.config/tanzu/tkg/unmanaged/boooooop/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster boooooop
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/josh/.config/tanzu/tkg/unmanaged/boooooop/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to boooooop

```

```
$ k get po -A | grep -i calico
kube-system          calico-kube-controllers-cb5567d94-pnk84          0/1     Pending    0          2s
kube-system          calico-node-g5d4p                                0/1     Init:0/3   0          2s
```

stmcginnis commented:
With `make install-all-tanzu-cli-plugins ENVS=darwin-amd64`

✅ cluster creation
```
tanzu um create choochoo

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   Downloading to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5.          Downloading to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5..         Downloaded to: /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/config.yaml
   Bootstrap Logs: /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster choochoo
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /Users/smcginnis/.config/tanzu/tkg/unmanaged/choochoo/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to choochoo

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete choochoo
```

✅ cluster configuration
```
❯ tanzu um configure -c calico chugchug
Wrote configuration file to: chugchug.yaml

❯ bat chugchug.yaml
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: chugchug.yaml
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ ClusterName: chugchug
   2   │ KubeconfigPath: ""
   3   │ NodeImage: ""
   4   │ Provider: kind
   5   │ ProviderConfiguration: {}
   6   │ Cni: calico
   7   │ CniConfiguration: {}
   8   │ PodCidr: 10.244.0.0/16
   9   │ ServiceCidr: 10.96.0.0/16
  10   │ TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
  11   │ PortsToForward: []
  12   │ SkipPreflight: false
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

✅ cluster creation with configuration (some minimal settings tweaked)
```
❯ tanzu um create -f chugchug.yaml

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /Users/smcginnis/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/config.yaml
   Bootstrap Logs: /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster chugchug
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /Users/smcginnis/.config/tanzu/tkg/unmanaged/chugchug/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to chugchug

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete chugchug
```

✅ cluster listing
```
❯ tanzu um ls
  NAME      PROVIDER
  choochoo  kind
  chugchug  kind
❯ tanzu um ls -o yaml
- name: choochoo
  provider: kind
- name: chugchug
  provider: kind
❯ tanzu um ls -o json
[
  {
    "name": "choochoo",
    "provider": "kind"
  },
  {
    "name": "chugchug",
    "provider": "kind"
  }
]
❯ tanzu um ls -o json | jq '.[] | .name'
"choochoo"
"chugchug"
```

✅ cluster deletion
```
❯ tanzu um rm chugchug

🧪 Deleting cluster: chugchug

🧪 Deleted cluster: chugchug
❯ tanzu um rm choochoo

🧪 Deleting cluster: choochoo

🧪 Deleted cluster: choochoo
```

jpmcb commented:
✅ cluster creation

```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 create meow

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/meow/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/meow/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster meow
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/meow/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   antrea.tanzu.vmware.com:0.13.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to meow

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete meow
```

cluster configuration
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 configure -c calico woofwoof
Wrote configuration file to: woofwoof.yaml
❯ cat woofwoof.yaml
ClusterName: woofwoof
KubeconfigPath: ""
NodeImage: ""
Provider: kind
ProviderConfiguration: {}
Cni: calico
CniConfiguration: {}
PodCidr: 10.244.0.0/16
ServiceCidr: 10.96.0.0/16
TkrLocation: projects.registry.vmware.com/tce/tkr:v1.21.5
PortsToForward: []
SkipPreflight: false

```

✅ cluster creation with configuration (some minimal settings tweaked)

```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 create -f woofwoof.yaml woof

📁 Created cluster directory

🔧 Resolving Tanzu Kubernetes Release (TKR)
   projects.registry.vmware.com/tce/tkr:v1.21.5
   TKR exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.21.5
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/woof/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/woof/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.21.5

📦 Selected core package repository
   projects.registry.vmware.com/tkg/packages/core/repo:v1.21.2_vmware.1-tkg.1

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.9.1

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tkg/packages/core/kapp-controller:v0.23.0_vmware.1-tkg.1

🚀 Creating cluster woof
   Base image downloaded
   Cluster created
   To troubleshoot, use:
   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/woof/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.tanzu.vmware.com:3.11.3+vmware.1-tkg.1

✅ Cluster created

🎮 kubectl context set to woof

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete woof
```

✅ cluster listing
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls
  NAME         PROVIDER
  hello-world  kind
  meow         kind
  meow01       kind
  woof         kind

❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o json
[
  {
    "name": "hello-world",
    "provider": "kind"
  },
  {
    "name": "meow",
    "provider": "kind"
  },
  {
    "name": "meow01",
    "provider": "kind"
  },
  {
    "name": "woof",
    "provider": "kind"
  }
]
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o yaml
- name: hello-world
  provider: kind
- name: meow
  provider: kind
- name: meow01
  provider: kind
- name: woof
  provider: kind
```

✅ cluster deletion
```
❯ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 ls -o json | jq '.[].name' | xargs -I _ ./artifacts/unmanaged-cluster/v0.10.0-fake.4/tanzu-unmanaged-cluster-linux_amd64 delete _

🧪 Deleting cluster: hello-world

🧪 Deleted cluster: hello-world

🧪 Deleting cluster: meow

🧪 Deleted cluster: meow

🧪 Deleting cluster: meow01

🧪 Deleted cluster: meow01

🧪 Deleting cluster: woof

🧪 Deleted cluster: woof

```

jpmcb reviewed (dismissed):
Let's do this 🚀

Commit:
sync package-release.sh with head
